addons/udes_sale_stock/tests/test_sale_order_state.py

```diff
@@ -0,0 +1,110 @@
+from . import common
+
+
+class TestSaleOrderState(common.BaseSaleUDES):
+    @classmethod
+    def setUpClass(cls):
+        super(TestSaleOrderState, cls).setUpClass()
+
+        Customer = cls.env["res.partner"]
+
+        cls.customer = Customer.create(
+            {
+                "name": "Bob",
+                "street": "1 Bobs House",
+                "street2": "Bobs street",
+                "city": "The city of Bob",
+                "zip": "EN1 4LS",
+            }
+        )
+
+        cls.quant = cls.create_quant(
+            cls.apple.id, cls.test_location_01.id, 30, package_id=cls.create_package().id
+        )
+
+        # Create sale order
+        sale = cls.create_sale(cls.customer)
+        cls.sale_line = cls.create_sale_line(sale, cls.apple, 15)
+        sale.requested_date = "2020-01-03"
+        sale.action_confirm()
+        cls.sale = sale
+
+        # Compute ordered list of pickings
+        ordered_pickings = []
+        picking = sale.picking_ids.filtered(lambda x: x.state == "assigned")
+        while picking:
+            ordered_pickings.append(picking)
+            picking = picking.u_next_picking_ids
+        cls.ordered_pickings = ordered_pickings
+
+        assert len(ordered_pickings) > 1
+        assert len(ordered_pickings) == len(sale.picking_ids)
+
+    def complete_picking(self, picking):
+        """
+        Completes a picking.
+        """
+        Location = self.env["stock.location"]
+
+        picking.ensure_one()
+
+        # Select a non-view location as the destination
+        dest_location = picking.location_dest_id
+        if dest_location.usage == "view":
+            dest_location = Location.search([
+                ("id", "child_of", picking.location_dest_id.id),
+                ("usage", "=", "internal"),
+            ])[0]
+
+        for ml in picking.move_line_ids:
+            ml.write({
+                "qty_done": ml.product_uom_qty,
+                "location_dest_id": dest_location.id,
+            })
+
+        picking.action_done()
+        self.assertEqual(picking.state, "done")
+
+    def test01_test_done(self):
+        """
+        Test that a sale order is done when all its pickings have been
+        completed.
+        """
+        # Complete the pickings in order
+        for picking in self.ordered_pickings:
+            self.assertEqual(self.sale.state, "sale")
+            self.assertFalse(self.sale_line.is_cancelled)
+
+            # Complete the picking
+            self.assertEqual(picking.state, "assigned")
+            self.complete_picking(picking)
+            self.assertEqual(picking.state, "done")
+
+        # Check that the sale order is done
+        self.assertEqual(self.sale.state, "done")
+        self.assertFalse(self.sale_line.is_cancelled)
+
+    def test02_test_last_picking_cancelled(self):
```

Comment:
Can we add another test with a sale order of two lines and cancel one line but not the second one?

Reply:
A new test has been added